### ISSUE TITLE
`draw_networkx_edges` does not adhere to `g.edges()` order when self-loops are present

### ISSUE THREAD

**Issue by mgbckr** [networkx/networkx#5106]
### Current Behavior
`draw_networkx_edges` seems to jumble formatting arguments for individual self-loop edges.  

### Expected Behavior
Among other example, in the code snippet below, the upper right self transition should be orange and the upper left one pink. Luckily the normal edges seem to be formatted fine.

### Steps to Reproduce
```
import networkx as nx
g = nx.Graph()
node_pos = {0:(0,0),1:(1,1),2:(0,1)}
g.add_edges_from([(0,0), (0,1), (1,1), (2,2), (2,1)])
print(g.edges())
nx.draw_networkx_edges(g, node_pos, edge_color=["red", "green", "orange", "blue", "pink"], width=[5,10,1,50,100])
```

### Environment
Python version: 3.9
NetworkX version: 2.6.3

### Additional context
![image](https://user-images.githubusercontent.com/1241516/135012412-3da601be-1506-49f0-ac1c-cc1b4abb1df5.png)

**Comment by dschult** [comment]
I can verify this color issue. After playing with it a little, it looks to me like the edges are drawn first with the correct colors and then the selfloops are drawn again using the colors in the order specified, but only applying them to the selfloops.

So we get `[(0,0): red, (0,1): green, (1,1): orange, (2,1): blue, (2,2): pink]`   
but then we redraw the self-loops:   `[(0,0): red, (1,1): green, (2,2): orange]` 
Sorry for the cryptic notation -- hope it is clear. 

We do draw the self-loops after all edges are drawn. So, it is possible that we need to adjust the handling of color there. I haven't dived into that code.

**Comment by rossbar** [comment]
You're spot-on @dschult - self-loops are handled separately from non-self edges (prior to 2.6, they were silently ignored and weren't drawn at all for undirected graphs). This breaks the expected 1-to-1 mapping between edges and `len(edges)` containers when there is a mixture of edges + self-loops. I expect this bug affects every kwarg where edge properties are specified in a `len(edgelist)` container, e.g. `width` in the above example, which I think exhibits the same problem: `(1, 1)` has width `10` instead of the desired `1`, `(2, 2)` has width `1` instead of the desired `50`.

**Comment by dschult** [comment]
Very helpful Issue @mgbckr !

Thanks @rossbar for explaining how it works and noticing that every kwarg representing edge properties probably is affected.  Would it be better to construct dicts instead of sequences where the dicts are keyed by edge tuple? The Matplotlib API seems to rely on arrays (or sequences) quite a lot. Presumably that makes a performance difference.

We are breaking that pattern though when we process some edges separately from others.  I suppose we could put the burden for separating onto the user. But maybe we could preprocess the self-loops and separating the self-loop info from the regular edge info.  But that might get messy...    

It's not clear to me what is the best path forward.

**Comment by rossbar** [comment]
Having looked at this again, I think the problem extends beyond the graph-with-selfloops case. There can be mapping problems when specifying a sequence of visualization properties (e.g. a list of edge colors) even with no self-loops are present. This arises from the fact that the default `edgelist=list(G.edges)` might have a different ordering than a user expects. For example:

```python
>>> G = nx.Graph([(0, 1), (2, 1), (3, 4), (2, 4)])
>>> edge_colors = ["red", "green", "blue", "black"]
>>> pos = nx.circular_layout(G)
>>> nx.draw_networkx_nodes(G, pos)
>>> nx.draw_networkx_labels(G, pos)
>>> nx.draw_networkx_edges(G, pos, edge_color=edge_colors)
```

I naively expected the `edge_colors` to map to the edges in the order they were specified in the `Graph` constructor (i.e. (0, 1)->red, (2, 1)->green, (3, 4)->blue and (2, 4)->black). However, the edgecolors of the last two edges are flipped:

![ed](https://user-images.githubusercontent.com/1268991/155791887-b388cd0f-fc6c-4905-8b61-590887819090.png)

This is due to the fact that `list(G.edges)` returns a different ordering than the edgelist used to create the graph:

```
>>> list(G.edges)
[(0, 1), (1, 2), (2, 4), (3, 4)]
```

So what this means is - the only way you can currently *guarantee* that a sequence of properties (colors, sizes, etc.) maps to the edges you expect is to also pass in `edgelist` where the ordering explicitly matches the order of the properties! (though the self-loops ordering would likely still be broken, at least for undirected graphs).

**Comment by rossbar** [comment]
This was mostly resolved by #5407. There's still the issue of the mapping of sequences of edge attributes to the correct edges, but that's a more involved issue that is more of a discussion topic than anything immediately actionable. Thus I'll close this as resolved. If related issue pop up, please open a new issue!
